docs/conf.py

```diff
@@ -385,53 +385,53 @@ def linkcode_resolve(domain, info):
+    except (AttributeError, ValueError):
+        # AttributeError:
+        # Object might be a non-static attribute of a class, e.g., self.num_qubits, which would only exist after init
+        # was called.
+        # For the moment we don't need a link for that as there is a link for the class already
+        #
+        # ValueError:
+        # info['module'] is empty
+        return None
+    try:
+        filepath = inspect.getsourcefile(obj)
+        line_number = inspect.getsourcelines(obj)[1]
+    except TypeError:
+        # obj might be a property or a static class variable, e.g.,
+        # loop_tag_id in which case obj is an int and inspect will fail
         try:
-            if 'module' in info and 'fullname' in info and info['module'] and info['fullname']:
-                obj = eval(info['module'] + '.' + info['fullname'])
+            # load obj one hierarchy higher (either class or module)
+            new_higher_name = info['fullname'].split('.')
+            module = importlib.import_module(info['module'])
+            if len(new_higher_name) > 1:
+                obj = module
             else:
-                return None
-        except AttributeError:
-            # Object might be a non-static attribute of a class, e.g.,
-            # self.num_qubits, which would only exist after init was called.
-            # For the moment we don't need a link for that as there is a link
-            # for the class already
-            return None
-        try:
+                obj = recursive_getattr(module, '.' + '.'.join(new_higher_name[:-1]))
+
             filepath = inspect.getsourcefile(obj)
             line_number = inspect.getsourcelines(obj)[1]
-        except:
-            # obj might be a property or a static class variable, e.g.,
-            # loop_tag_id in which case obj is an int and inspect will fail
-            try:
-                # load obj one hierarchy higher (either class or module)
-                new_higher_name = info['fullname'].split('.')
-                if len(new_higher_name) <= 1:
-                    obj = eval(info['module'])
-                else:
-                    obj = eval(info['module'] + '.' + '.'.join(new_higher_name[:-1]))
-                filepath = inspect.getsourcefile(obj)
-                line_number = inspect.getsourcelines(obj)[1]
-            except:
-                return None
-        # Only require relative path projectq/relative_path
-        projectq_path = inspect.getsourcefile(projectq)[:-11]
-        relative_path = os.path.relpath(filepath, projectq_path)
-        url = github_url + github_tag + "/projectq/" + relative_path + "#L" + str(line_number)
-        return url
+        except AttributeError:
+            return None
 
+    # Only require relative path projectq/relative_path
+    projectq_path = os.path.realpath(os.path.join(os.path.dirname(__file__), '..')) + os.path.sep
+    idx = len(projectq_path)
+    relative_path = filepath[idx:]
```

Comment:
So essentially, we want to override the normal link generation made by Sphinx (which by default does nothing I believe) in order so that the generated documentation links to the GitHub repository.

Now, in order to generate the proper link, we need to compute the relative path from the root directory.

Line 424 computes the absolute path of the root directory (with a path separator added to the end).
Line 425-426 then removes the absolute path to the root directory from the file path so that we are left with the relative path of the file under consideration.

All of this works since the path returned by `inspect.getsourcefile` is guaranteed to be absolute since we all the files we consider are not in the current python module.

Reply:
ah, sorry. My bad. Thanks, for explaining. Now I see it. I misread `filepath` in line 426 as `projectq_path` and that got me confused.